Solution1/BusinessLayer/Services/WeatherService.cs

```diff
@@ -54,5 +57,53 @@ public async Task<ForecastWeatherDTO> GetForecastByCityNameAsync(string cityName
 
             return _mapper.Map<ForecastWeatherDTO>(forecast).FillCommentByTemp(); 
         }
+
+        public async Task<Dictionary<bool, IEnumerable<WeatherResponseDTO>>> GetWeatherByArrayCityNameAsync(IEnumerable<string> cityNames)
+        {
+            var timer = new Stopwatch();
+            timer.Start();
+
+            var listTasksRequest = cityNames.Select(async cityName =>
+            {
+                var weatherResponseDTO = new WeatherResponseDTO() { CityName = cityName };
+                try
+                {
+                    ValidationCityName(cityName);
+                    var temp = (await _weatherApiService.GetByCityNameAsync(cityName))?.TemperatureValues.Temp;
+                    if (temp.HasValue)
+                    {
+                        weatherResponseDTO.Temp = temp.Value;
+                        weatherResponseDTO.IsSuccessfulRequest = true;
+                    }
+                }
+                catch (ValidationException ex)
+                {
+                    weatherResponseDTO.IsSuccessfulRequest = false;
+                    weatherResponseDTO.ErrorMessage = ex.Errors.FirstOrDefault().ErrorMessage;
+                }
+                catch (Exception ex)
+                {
+                    weatherResponseDTO.IsSuccessfulRequest = false;
+                    weatherResponseDTO.ErrorMessage = ex.Message;
+                }
+                weatherResponseDTO.LeadTime = timer.ElapsedMilliseconds;
+                return weatherResponseDTO;
+            });
+
+            var weatherResponses = await Task.WhenAll(listTasksRequest);
+            return weatherResponses.GroupBy(w => w.IsSuccessfulRequest).ToDictionary(k => k.Key, v => v.AsEnumerable());
```

Comment:
why asenumerable? if array returned, this means it's already calculated, so not necessary to do make it enumerable back - so that it will be calculated after some time again

Reply:
+